<a href="https://colab.research.google.com/github/HagarIbrahiem/Learning_purposes/blob/main/365__Feature_Engineering__Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt

In [2]:
from google.colab import drive
drive.mount('/content/drive')
!ls drive/'My Drive'/ML_Process_Data_Files/Section_7_Feature_Engineering
df = pd.read_csv("drive/My Drive/ML_Process_Data_Files/Section_7_Feature_Engineering/airbnb_dataset_ml_process.csv")

Mounted at /content/drive
AB_NYC_2019.csv  airbnb_dataset_ml_process.csv	vehicles_sample.csv


In [3]:
df.head(1)

Unnamed: 0    id                                name  host_id host_name  \
0           0  2539  Clean & quiet apt home by the park     2787      John   

  neighbourhood_group neighbourhood  latitude  longitude     room_type  price  \
0            Brooklyn    Kensington  40.64749  -73.97237  Private room    149   

   minimum_nights  number_of_reviews last_review  reviews_per_month  \
0               1                  9  2018-10-19               0.21   

   calculated_host_listings_count  availability_365      expensive  
0                               6               365  non-expensive

##One-hot encoding

In [6]:
df['expensive'].value_counts()

non-expensive    36718
expensive        12177
Name: expensive, dtype: int64

In [9]:
##knowing average price of Expensive category
df.groupby(['expensive']).mean()['price']

## so avergae price of (expensive) category is 335
## so avergae price of (non-expensive) category is 92


expensive
expensive        335.295968
non-expensive     92.172204
Name: price, dtype: float64

In [10]:
dummies = pd.get_dummies(df['expensive'])
dummies.head()

expensive  non-expensive
0          0              1
1          1              0
2          0              1
3          0              1
4          0              1

**Multi-Hot Encodin**

In [12]:
df['neighbourhood'].unique()

array(['Kensington', 'Midtown', 'Harlem', 'Clinton Hill', 'East Harlem',
       'Murray Hill', 'Bedford-Stuyvesant', "Hell's Kitchen",
       'Upper West Side', 'Chinatown', 'South Slope', 'West Village',
       'Williamsburg', 'Fort Greene', 'Chelsea', 'Crown Heights',
       'Park Slope', 'Windsor Terrace', 'Inwood', 'East Village',
       'Greenpoint', 'Bushwick', 'Flatbush', 'Lower East Side',
       'Prospect-Lefferts Gardens', 'Long Island City', 'Kips Bay',
       'SoHo', 'Upper East Side', 'Prospect Heights',
       'Washington Heights', 'Woodside', 'Brooklyn Heights',
       'Carroll Gardens', 'Gowanus', 'Flatlands', 'Cobble Hill',
       'Flushing', 'Boerum Hill', 'Sunnyside', 'DUMBO', 'St. George',
       'Highbridge', 'Financial District', 'Ridgewood',
       'Morningside Heights', 'Jamaica', 'Middle Village', 'NoHo',
       'Ditmars Steinway', 'Flatiron District', 'Roosevelt Island',
       'Greenwich Village', 'Little Italy', 'East Flatbush',
       'Tompkinsville', 'Asto

In [14]:
dummies_neighbourhood = pd.get_dummies(df['neighbourhood'])
dummies_neighbourhood.head()

Allerton  Arden Heights  Arrochar  Arverne  Astoria  Bath Beach  \
0         0              0         0        0        0           0   
1         0              0         0        0        0           0   
2         0              0         0        0        0           0   
3         0              0         0        0        0           0   
4         0              0         0        0        0           0   

   Battery Park City  Bay Ridge  Bay Terrace  Bay Terrace, Staten Island  ...  \
0                  0          0            0                           0  ...   
1                  0          0            0                           0  ...   
2                  0          0            0                           0  ...   
3                  0          0            0                           0  ...   
4                  0          0            0                           0  ...   

   Westerleigh  Whitestone  Williamsbridge  Williamsburg  Willowbrook  \
0            0           0               0             0            0   
1            0           0               0             0            0   
2            0           0               0             0            0   
3            0           0               0             0            0   
4            0           0               0             0            0   

   Windsor Terrace  Woodhaven  Woodlawn  Woodrow  Woodside  
0                0          0         0        0         0  
1                0          0         0        0         0  
2                0          0         0        0         0  
3                0          0         0        0         0  
4                0          0         0        0         0  

[5 rows x 221 columns]

Now we have 221 Columns encoded , this is called **high cardinality**, resulting in:
- overfitting, 
- large memory consumption  
- slow training times

##2- Ordinal Encoding

There are a number of solutions to solving the cardinality problem. The first one is ordinal encoding. Ordinal encoding is a method of replacing categories with numbers. However, these numbers have an inherent ordering to them

In [17]:
df['room_type'].value_counts()

Entire home/apt    25409
Private room       22326
Shared room         1160
Name: room_type, dtype: int64

In [23]:
from sklearn.preprocessing import OrdinalEncoder

_OrdinalEncoder = OrdinalEncoder()
result = _OrdinalEncoder.fit_transform(df[['room_type']])
result = pd.DataFrame(result)
result.head()



0.0    25409
1.0    22326
2.0     1160
dtype: int64


In [26]:
print(result.value_counts())

room_type_ord_encoded
0.0                      25409
1.0                      22326
2.0                       1160
dtype: int64


In [25]:
## Aggregate it back into original dataframe
ord_encoded = pd.DataFrame(result)
ord_encoded.columns = ['room_type_ord_encoded']

ord_encoded_df = pd.concat([df, ord_encoded], axis = 1)



Unnamed: 0    id                                              name  \
0           0  2539                Clean & quiet apt home by the park   
1           1  2595                             Skylit Midtown Castle   
2           2  3647               THE VILLAGE OF HARLEM....NEW YORK !   
3           3  3831                   Cozy Entire Floor of Brownstone   
4           4  5022  Entire Apt: Spacious Studio/Loft by central park   

   host_id    host_name neighbourhood_group neighbourhood  latitude  \
0     2787         John            Brooklyn    Kensington  40.64749   
1     2845     Jennifer           Manhattan       Midtown  40.75362   
2     4632    Elisabeth           Manhattan        Harlem  40.80902   
3     4869  LisaRoxanne            Brooklyn  Clinton Hill  40.68514   
4     7192        Laura           Manhattan   East Harlem  40.79851   

   longitude        room_type  price  minimum_nights  number_of_reviews  \
0  -73.97237     Private room    149               1                  9   
1  -73.98377  Entire home/apt    225               1                 45   
2  -73.94190     Private room    150               3                  0   
3  -73.95976  Entire home/apt     89               1                270   
4  -73.94399  Entire home/apt     80              10                  9   

  last_review  reviews_per_month  calculated_host_listings_count  \
0  2018-10-19               0.21                               6   
1  2019-05-21               0.38                               2   
2         NaN                NaN                               1   
3  2019-07-05               4.64                               1   
4  2018-11-19               0.10                               1   

   availability_365      expensive  room_type_ord_encoded  
0               365  non-expensive                    1.0  
1               355      expensive                    0.0  
2               365  non-expensive                    1.0  
3               194  non-expensive                    0.0  
4                 0  non-expensive                    0.0

In [45]:
ord_encoded_df[['room_type','room_type_ord_encoded'] ].head()


room_type  room_type_ord_encoded
0     Private room                    1.0
1  Entire home/apt                    0.0
2     Private room                    1.0
3  Entire home/apt                    0.0
4  Entire home/apt                    0.0

##3- Frequency Encoding

Another technique to address the cardinality issue is frequency encoding. Rather than replace our categorical variables with ordinal variables, we're going to replace our categories with the frequency in which they occur.

In [33]:
frequency_encoded= df['neighbourhood'].value_counts()
frequency_encoded


Williamsburg          3920
Bedford-Stuyvesant    3714
Harlem                2658
Bushwick              2465
Upper West Side       1971
                      ... 
Fort Wadsworth           1
Richmondtown             1
New Dorp                 1
Rossville                1
Willowbrook              1
Name: neighbourhood, Length: 221, dtype: int64

In [35]:
frequency_encoded['Allerton']

42

In [36]:
df['neighbourhood'].apply(lambda x : frequency_encoded[x])

0         175
1        1545
2        2658
3         572
4        1117
         ... 
48890    3714
48891    2465
48892    2658
48893    1958
48894    1958
Name: neighbourhood, Length: 48895, dtype: int64

In [38]:
## Implmenting FrequencyEncoder Class
class FrequencyEncoder:
    def fit(self, train_df, column):
        self.train_df = train_df
        self.column = column
        
    def transform(self, test_df, column):
        frequency_encoded = self.train_df.groupby([self.column]).size()
        col_name = column + '_freq'
        test_df.loc[:,col_name] = test_df[column].apply(lambda x: frequency_encoded[x])
        return test_df

# frequency_encoding(df, column='neighbourhood')

_FrequencyEncoder = FrequencyEncoder()
_FrequencyEncoder.fit(df, column='neighbourhood')
df_freq_enc = _FrequencyEncoder.transform(df, column='neighbourhood')

df_freq_enc = pd.DataFrame ( df_freq_enc)

In [44]:
df_freq_enc[['neighbourhood_group','neighbourhood_freq'] ].head()

neighbourhood_group  neighbourhood_freq
0            Brooklyn                 175
1           Manhattan                1545
2           Manhattan                2658
3            Brooklyn                 572
4           Manhattan                1117

##4- Target Encoding

Another technique to address the cardinality issue is Target encoding. While at frequency encoding, we encode the number of instances a category occurs as its value, we here encode the mean of our target variable per each category 

In [46]:
df.groupby('neighbourhood').mean()['price']

neighbourhood
Allerton            87.595238
Arden Heights       67.250000
Arrochar           115.000000
Arverne            171.779221
Astoria            117.187778
                      ...    
Windsor Terrace    138.993631
Woodhaven           67.170455
Woodlawn            60.090909
Woodrow            700.000000
Woodside            85.097872
Name: price, Length: 221, dtype: float64

In [48]:
##implemnation of TargetEncoder Class 

class TargetEncoder:
    def fit(self, train_df, target_col, categ_col):
        self.train_df = train_df
        self.target_col = target_col
        self.categ_col = categ_col
        
    def transform(self, test_df, column = None):
        if column is None:
            column = self.categ_col
        
        target_encoder = self.train_df.groupby([self.categ_col]).mean()[self.target_col]

        df[self.categ_col].apply(lambda x: target_encoder[x])

        col_name = column + '_target_enc'
        test_df.loc[:,col_name] = test_df[column].apply(lambda x: target_encoder[x])
        return test_df
    
_TargetEncoder = TargetEncoder()
_TargetEncoder.fit(df, target_col = 'price', categ_col = 'neighbourhood')

te_df = _TargetEncoder.transform(df)



In [49]:
te_df[['neighbourhood','neighbourhood_target_enc']].head()

neighbourhood  neighbourhood_target_enc
0    Kensington                 92.885714
1       Midtown                282.719094
2        Harlem                118.974041
3  Clinton Hill                181.893357
4   East Harlem                133.198747

##5- Probability Ratio Encoding

Probability Ratio Encoding is similar to target encoding. But rather than using the mean of the target, we're looking at the probability this category is going to be a positive label.

**P.S.** Probability Ratio Encoding only used when the target variable is **Categorical** means **classification** problem

In [50]:
## implemnt ProbabilityRatioEncoder Class
class ProbabilityRatioEncoder:
    def fit(self, train_df, categ_col, target_col):
        self.train_df = train_df
        self.categ_col = categ_col
        self.target_col = target_col
        
    def transform(self, test_df, constant = 0):
        totals = self.train_df.groupby([self.categ_col]).size() 
        sums = self.train_df.groupby([self.categ_col]).sum()[self.target_col]

        ratio_encoder = (sums+ constant)/totals   ## constant so we can divide 0/totals , thus to have infinity
        
        col_name = self.categ_col + '_prob_ratio'
        test_df.loc[:,col_name] = test_df[self.categ_col].apply(lambda x: ratio_encoder[x])
        return test_df

In [52]:
## creating a categorical target variable (expensive) becuase (price) target variable is numeric. 
## breake new target variable to positive and negative
target_variable = pd.get_dummies(df['expensive'])

df = pd.concat([df,target_variable], axis = 1)

In [55]:
df.head(1)

Unnamed: 0    id                                name  host_id host_name  \
0           0  2539  Clean & quiet apt home by the park     2787      John   

  neighbourhood_group neighbourhood  latitude  longitude     room_type  ...  \
0            Brooklyn    Kensington  40.64749  -73.97237  Private room  ...   

   number_of_reviews  last_review  reviews_per_month  \
0                  9   2018-10-19               0.21   

  calculated_host_listings_count  availability_365      expensive  \
0                              6               365  non-expensive   

   neighbourhood_freq neighbourhood_target_enc  expensive  non-expensive  
0                 175                92.885714          0              1  

[1 rows x 22 columns]

In [56]:
_ProbabilityRatioEncoder = ProbabilityRatioEncoder()

_ProbabilityRatioEncoder.fit(df, 'neighbourhood','expensive')
pre_df = _ProbabilityRatioEncoder.transform(df)

In [59]:
pre_df[['neighbourhood','expensive','non-expensive','neighbourhood_prob_ratio']].head()

neighbourhood      expensive  expensive  non-expensive  \
0    Kensington  non-expensive          0              1   
1       Midtown      expensive          1              0   
2        Harlem  non-expensive          0              1   
3  Clinton Hill  non-expensive          0              1   
4   East Harlem  non-expensive          0              1   

   neighbourhood_prob_ratio  
0                  0.091429  
1                  0.623948  
2                  0.132807  
3                  0.243007  
4                  0.168308

##6- Weight of Evidence Encoding

Weight of Evidence encoding is similar to probability ratio encoding. The only difference, is we're applying a log transform on top of the probability ratio transformation:

In [60]:
## implemnt WeightofEvidenceEncoder class
class WeightofEvidenceEncoder:
    def fit(self, train_df, categ_col, target_col):
        self.train_df = train_df
        self.categ_col = categ_col
        self.target_col = target_col
        
    def transform(self, test_df, constant = 0):
        totals = self.train_df.groupby([self.categ_col]).size() 
        sums = self.train_df.groupby([self.categ_col]).sum()[self.target_col]

        woe_encoder = np.log((sums+ constant)/totals)
        
        col_name = self.categ_col + '_woe'
        test_df.loc[:,col_name] = test_df[self.categ_col].apply(lambda x: woe_encoder[x])
        return test_df
    
woe = WeightofEvidenceEncoder()

woe.fit(df, 'neighbourhood','expensive')
woe_df = woe.transform(df)


/usr/local/lib/python3.8/dist-packages/pandas/core/arraylike.py:364: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


In [61]:
woe_df[['neighbourhood','expensive','non-expensive','neighbourhood_woe']].head()

neighbourhood      expensive  expensive  non-expensive  neighbourhood_woe
0    Kensington  non-expensive          0              1          -2.392197
1       Midtown      expensive          1              0          -0.471688
2        Harlem  non-expensive          0              1          -2.018861
3  Clinton Hill  non-expensive          0              1          -1.414665
4   East Harlem  non-expensive          0              1          -1.781960

##7-Binning

The last technique is called binning. This is where we take a continuous variable and bin them into different buckets, thus, transforming this continuous variable into a categorical variable:

In [68]:
df['number_of_reviews'].value_counts()

0      10052
1       5244
2       3465
3       2520
4       1994
       ...  
313        1
540        1
480        1
326        1
341        1
Name: number_of_reviews, Length: 394, dtype: int64

In [69]:
def reviews_bin(x):
    if x < 50:
        return 'less_50'
    if x >= 50 and x < 100:
        return '50_to_100'
    if x >= 100 and x < 150:
        return '100_to_150'
    if x >= 150 and x < 200:
        return '150_to_200'
    if x >= 200:
        return '200_plus'
    
df['number_of_reviews'].map(reviews_bin)

0         less_50
1         less_50
2         less_50
3        200_plus
4         less_50
           ...   
48890     less_50
48891     less_50
48892     less_50
48893     less_50
48894     less_50
Name: number_of_reviews, Length: 48895, dtype: object

##Modeling + Comparison

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split


def run_logistic_regression_pipeline(df):
    
    df.replace([np.inf, -np.inf], np.nan, inplace=True)
    df = df.dropna()

    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    X = df.select_dtypes(include=numerics)
    y = df['price']

    X_train = X[:30000]
    y_train = y[:30000]
    X_test = X[30000:]
    y_test = y[30000:]

    lr = LogisticRegression()
    lr.fit(X_train, y_train)
    preds = lr.predict(X_test)

    accuracy = accuracy_score(preds, y_test)
    
    return accuracy

# Hot Encoding
hot_encode_accuracy_score = run_logistic_regression_pipeline(hot_encoded_df)

# ## Ordinal Encoding
ordinal_accuracy_score = run_logistic_regression_pipeline(ord_encoded_df)

# ## Target Encoding
target_encode_accuracy_score = run_logistic_regression_pipeline(te_df)

# ## Probability Ratio Encoding
prob_ratio_accuracy_score = run_logistic_regression_pipeline(pre_df)

# ## Weight of Evidence Encoding
# woe_accuracy_score = run_logistic_regression_pipeline(woe)



In [ ]:

print('Hot Encoded Accuracy Score: %.3f' % hot_encode_accuracy_score)
print('Ordinal Encoded Accuracy Score: %.3f' % ordinal_accuracy_score)
print('Target Encoded Accuracy Score: %.3f' % target_encode_accuracy_score)
print('Probability Ratio Encoded Accuracy Score: %.3f' % prob_ratio_accuracy_score)
# print('Weight of Evidence Encoded Accuracy Score: %.3f' % woe_accuracy_score)